src/collapse/common.rs

```diff
@@ -0,0 +1,780 @@
+                if n == 0 {
+                    // If we've reached the end of the data, send the final chunk to the worker
+                    // threads and break from the loop, The worker threads may or may not still
+                    // be alive (depending on if one errored in between the sending of the last
+                    // chunk and the sending of this one), but either way we should break the loop;
+                    // so there's no need to check for a `SendError` here.
+                    let _ = tx_input.send(Some(buf));
+                    break;
+                }
+                let line = &buf[index..index + n];
+                index += n;
+                if self.is_end_of_stack(line) {
+                    // If we've reached the end of a stack, count it.
+                    nstacks += 1;
+                    if nstacks == nstacks_per_job {
+                        // If we've accumulated enough stacks to make up a chunk to send to the
+                        // worker threads, try to send it.
+                        let buf_capacity = usize::next_power_of_two(buf.capacity());
+                        let chunk = mem::replace(&mut buf, Vec::with_capacity(buf_capacity));
+                        if tx_input.send(Some(chunk)).is_err() {
+                            // If sending the chunk produces a `SendError`, this means that one
+                            // of the worker threads has errored, sent a signal to all the other
+                            // worker threads to shut down, and they have all shutdown, in which
+                            // case we know there will be an error waiting for us on the error
+                            // channel; so we should stop parsing input data (i.e. break).
+                            break;
+                        }
+                        index = 0;
+                        nstacks = 0;
+                    }
+                    continue
+                }
+            }
+
+            // We've run out of input data to send to the worker threads; so tell them.
+            for _ in &handles {
+                if tx_input.send(None).is_err() {
+                    // If we're unable to send data on the input channel, it means one of the
+                    // worker threads has errored and signaled to all the other worker threads
+                    // to shutdown (and they all have), in which case, as above, we should stop
+                    // trying to send information to them and proceed directly to the checking
+                    // the error channel for the error that will be there.
+                    break;
+                }
+            }
+
+            // The main thread needs to drop it's handle to the error sender here because we
+            // are about to poll the error receiver for errors, which will block until all
+            // the error senders have been dropped (including ours).
+            drop(tx_error);
```

Comment:
I think we can drop this immediately after we start the workers?

Reply:
Sure, either location would do the same thing. I put it right before checking the error channel, however, because in that location/context, it's easier to understand the comment about why we're dropping it. If you prefer dropping it immediately after starting the workers, that would be fine, but I'm (only very slightly) inclined to just leave it where it is.